### ISSUE TITLE
Differentiable Programming

### ISSUE THREAD

**Issue by allenleein** [allenleein/brains#4]
“Deep Learning has outlived its usefulness as a buzz-phrase“ (facebook.com)

[Text from post]
OK, Deep Learning has outlived its usefulness as a buzz-phrase. Deep Learning est mort. Vive Differentiable Programming!

Yeah, Differentiable Programming is little more than a rebranding of the modern collection Deep Learning techniques, the same way Deep Learning was a rebranding of the modern incarnations of neural nets with more than two layers.

But the important point is that people are now building a new kind of software by assembling networks of parameterized functional blocks and by training them from examples using some form of gradient-based optimization.

An increasingly large number of people are defining the network procedurally in a data-dependant way (with loops and conditionals), allowing them to change dynamically as a function of the input data fed to them. It's really very much like a regular progam, except it's parameterized, automatically differentiated, and trainable/optimizable. Dynamic networks have become increasingly popular (particularly for NLP), thanks to deep learning frameworks that can handle them such as PyTorch and Chainer (note: our old deep learning framework Lush could handle a particular kind of dynamic nets called Graph Transformer Networks, back in 1994. It was needed for text recognition).

People are now actively working on compilers for imperative differentiable programming languages. This is a very exciting avenue for the development of learning-based AI.

Important note: this won't be sufficient to take us to "true" AI. Other concepts will be needed for that, such as what I used to call predictive learning and now decided to call Imputative Learning. More on this later....

[HN](https://news.ycombinator.com/item?id=16082860)
[MIT Media lab](https://www.edge.org/response-detail/26794)

Question: 

FP in AI?

**Comment by allenleein** [comment]
Tangent is a new, free, and open-source Python library for automatic differentiation.

[Link](https://github.com/google/tangent)
